keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+`Node.Status.IPs` and `Node.Status.Hostname` by
+cloud-controller-manager.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+```
+<<[UNRESOLVED inconsistency ]>>
+Security issues related to Node.Status.Addresses and Node.Status.IPs
+containing different values?
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+### Implementation
+
+#### Hostname
+
+With the deprecation of `Node.Status.Addresses`, this value will be
+moved to `Node.Status.Hostname`, but does not otherwise need to be
+changed.
+
+```
+<<[UNRESOLVED hostname-label ]>>
+Is it a bug that `Node.Status.Hostname`/`NodeHostName` and the
+`"kubernetes.io/hostname"` label don't necessarily match? (NB: This
+is an existing behavior.)
+<<[/UNRESOLVED]>>
+```
+
+#### IPs
+
+The other information in `Node.Status.Addresses` will be moved to a
+new `Node.Status.IPs` array:
+
+    type NodeStatus struct {
+            ...
+
+            // List of the node's IPs. The first IP in the list is the "primary" node IP
+            // (eg, the hostIP value of pods on that node). If the second IP is of the
+            // opposite IP address family from the first (eg, IPv6 vs IPv4), then it is
+            // considered the primary IP of that address family. Other than that, the
+            // order of elements carries no meaning. Note that there may be a mix of IPv4
+            // and IPv6 addresses here regardless of whether the cluster supports both
+            // address families.
+            IPs []NodeIP `json:"ips"`
+
+            ...
+    }
+
+(Unlike with `Addresses`, this has no declared `patchStrategy` and thus
+defaults to `replace`, which is correct since it is solely maintained by
+a single component (kubelet or cloud-controller-manager), which always
+updates the entire field at once.)
+
+```
+<<[UNRESOLVED primary-ip ]>>
+If we're tightening up the rules on how cloud providers set addresses
+then it seems to make sense to just require them to put the primary IP
+first (as the comment above does) rather than having some complicated
+"first Internal IP unless there are no Internal IPs in which case..."
+rule like we have now.
+
+Instead of "If the second IP is of the opposite address family from
+the first, then it is considered the primary IP of that address
+family" we could instead do "If there are any IPs of the opposite
+address family from the first IP, then the first such IP is considered
+the primary IP of that address family" (which would, eg, let you put
+the IPv4 External IPs before the IPv6 Internal IPs if you wanted,
+etc). It seems like that just makes life more complicated for everyone
```

Comment:
I disagree. Assuming someone has read the detailed explanation seems more complicated. First occurrence per family works if you want them 1st and 2nd but also works if you want to track family groups together.

Reply:
I was thinking more from the perspective of code (ie, writing something to find the primary IPv4 and IPv6 IPs of a node) though after writing it out, it's (a) not _that_ terrible in the case where they can appear anywhere, and (b) not _that_ simple in the case where they have to be 1st and 2nd. And maybe the answer is just to have a few good helper methods anyway.